plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -506,13 +508,53 @@ public FlowExecutionOwner getOwner() {
     private TimingFlowNodeStorage createStorage() throws IOException {
         FlowNodeStorage wrappedStorage;
 
-        FlowDurabilityHint hint = getDurabilityHint();
-        wrappedStorage = (hint.isPersistWithEveryStep())
-                ? new SimpleXStreamFlowNodeStorage(this, getStorageDir())
-                : new BulkFlowNodeStorage(this, getStorageDir());
+        if (this.storageDir != null && this.storageDir.endsWith("-completed")) {
+           wrappedStorage = new BulkFlowNodeStorage(this, getStorageDir());
+        } else {
+            FlowDurabilityHint hint = getDurabilityHint();
+            wrappedStorage = (hint.isPersistWithEveryStep())
+                    ? new SimpleXStreamFlowNodeStorage(this, getStorageDir())
+                    : new BulkFlowNodeStorage(this, getStorageDir());
+        }
         return new TimingFlowNodeStorage(wrappedStorage);
     }
 
+    /**
+     * Called when the build completes to migrate from {@link SimpleXStreamFlowNodeStorage} to
+     * {@link BulkFlowNodeStorage} to improve read performance for completed builds.
+     */
+    private void optimizeStorage(FlowNode flowEndNode) {
+        if (storage.delegate instanceof SimpleXStreamFlowNodeStorage) {
+            String newStorageDir = (this.storageDir != null) ? this.storageDir + "-completed" : "workflow-completed";
```

Comment:
Similar to what we do with `workflow-fallback` in `createPlaceholderNodes`, but of course other approaches could be used.